### ISSUE TITLE
Don't run portable tests from NUnit Console

### ISSUE THREAD

**Issue by rprouse** [nunit/nunit#988]
Spawned from #948

We don't want tons of issues reported when users try to run tests compiled against the portable library and it fails to run. Better to have the driver look at the nunit.framework.dll and determine if it is portable before running it. If it is portable, give a message that portable tests must be run with platform specific runners or nunitlite.

**Comment by rprouse** [comment]
@CharliePoole, I was going to add the following to the framework assemblies, then check in the drivers;

```C#
#if PORTABLE
[assembly: AssemblyMetadata("PCL", "True")]
#else
[assembly: AssemblyMetadata("PCL", "False")]
#endif
```

I am running into Mono.Cecil problems though. In `DriverService.GetDriver`, we get the assembly references, but they are lightweight `AssemblyNameReference` objects. From there, I can load the referenced assembly in Cecil, then get the assembly attributes as a collection of `Mono.Cecil.CustomAttribute`.

The problem is that the API assembly does not reference Mono.Cecil, so I cannot pass in the custom attributes to make a decision based on them. I want it to be as generic as possible, so I want to pass in as much info as possible.

Possible solutions,

1. Wrap the custom attributes in a new class in the API assembly.
2. Add a reference to Mono.Cecil in the API assembly.

Both seem a bit heavy handed for an RC change. What do you think? Is there an easier way we could identify PCL assemblies that I am not thinking of?

**Comment by CharliePoole** [comment]
I'm a little shaky on why you need to have  the attributes defined in the API (which I agree seems a bit heavy-handed). Can't the driver service make the decision by itself? Or are you trying to think ahead for when we have a driver that will handle the PCL?

I see two (I think) simple ways to go:
1. If the service finds this attribute it simply returns the null driver (I forget it's name now) with a tailored message.
2. Go ahead and call the NUnit3 driver, which then finds the attribute using normal reflection and returns an error from it's Load entry point.

Am I missing something?

BTW, I'm not looking at the code right now - I can look a bit later.

**Comment by rprouse** [comment]
The attribute is defined on the framework to tag the framework as portable or not. Ideally, I would like to pass as much information as possible to each driver factory to allow them to decide if they can execute tests.

**Comment by CharliePoole** [comment]
Right. But the driver factory has access to the entire framework assembly. it can look at the attribute itself. Maybe I'm missing something though. Can you write a code snippet that shows what you're thinking of doing?

**Comment by rprouse** [comment]
The `IDriverFactory` is only passed in the assemblyName and the version of the assembly. These are passed in from `DriverService.GetDriver` which currently uses Mono.Cecil to read the assembly under test, then get the referenced assemblies as `Mono.Cecil.AssemblyNameReference` objects which are lightweight objects that contain minimal information about the referenced assemblies.

I don't want to pass Mono.Cecil objects into the drivers or the factories. I assume that factories will be extensions in the future and I don't want to force those extensions to take a dependency on Mono.Cecil. I could, but it feels wrong.

The other alternative is to pass a class defined in the API assembly that wraps the assembly and it's attributes and gives the drivers enough information to decide if they can run tests for that assembly.

The code that I am looking to change is in DriverService.GetDriver, but I am not sure what to pass in to `IsSupportedTestFramework` without loading assemblies using reflections which I thought we were avoiding with Mono.Cecil.

```C#
var references = AssemblyDefinition.ReadAssembly(assemblyPath).MainModule.AssemblyReferences;

foreach (var factory in _factories)
{
    foreach (var reference in references)
    {
        if (factory.IsSupportedTestFramework(reference.Name, reference.Version))
            return factory.GetDriver(domain, reference.Name, reference.Version);
    }
}
```

**Comment by CharliePoole** [comment]
I think we will eventually add to this interface or define an IDriverFactory2 interface. But we probably don't want to be changing our key interfaces in the middle of a release.

So for now, what if we did...

```
public bool IsSupportedTestFramework
{
    if (assemblyName != NUNIT_FRAMEWORK || version.Major < 3)
        return false;

    // Use .NET reflection to load the assembly and look at the attribute and set isPCL

    return !isPCL
}
```

There is no loss in doing the reflection load, because we are going to to be loading the assembly anyway in order to run tests.

Or - if you want to avoid checking other drivers  - just throw an NUnitEngineException if it's PCL. That will can be caught by DriverService and a NotRunnableFrameworkDriver should be returned.

**Comment by CharliePoole** [comment]
Here is a rather hacky solution. :-)

I replaced the call to CreateObject in NUnit3FrameworkDriver.Load with the following:

try
{
    _frameworkController = CreateObject(CONTROLLER_TYPE, testAssemblyPath, idPrefix, (System.Collections.IDictionary)settings);
}
catch (System.Runtime.Serialization.SerializationException ex)
{
    throw new NUnitEngineException("The NUnit 3.0 driver does not support the portable build of the framework.", ex);
}

That gives me a nice message.

Of course, it assumes that there is no other reason to get a SerializationException, which is not true even though unlikely and it doesn't allow for some other driver to take on the job of handling PCLs.

**Comment by skylerberg** [comment]
I think I ran into another case where a SerializationException in another case, which made this error message uninformative.

I edited and renamed a lot of files programmatically, with the side effect that a newline was added to the end of every file, including DLLs, which corrupted them. Now, I have corrected all the issues I could find and everything appears to work, except for my tests. When I run my tests, I get the error message introduced by this change.

If I manage to find the actual cause of the problem, then I will open an issue with more details.
